### ISSUE TITLE
MissingWhenCase false negative with nulls

### ISSUE THREAD

**Issue by BraisGabin** [detekt/detekt#3189]
## Expected Behavior
`MissingWhenCase` should check for null case if the enum or sealed can be `null` 

## Observed Behavior
It doesn't check that.

## Steps to Reproduce
```kotlin
fun whenOnEnumPass(c: Color?) {
    when(c) {
        Color.BLUE -> {}
        Color.GREEN -> {}
        Color.RED -> {}
    }
}
```

## Your Environment
* Version of detekt used: `master` (`1.14.2`)

**Comment by schalkms** [comment]
Closed due to the arguments made in the referenced PR.
- https://github.com/detekt/detekt/pull/3194#issuecomment-720625803
- https://github.com/detekt/detekt/pull/3194#issuecomment-720635060

**Comment by BraisGabin** [comment]
Should we create a tag "require smart cast" or something similar to track this issues? I mean, even if we close it to find them if we eventually implement it. Or if we want ro evaluate if it's worth it to implement it.

**Comment by schalkms** [comment]
An available smart cast detection alone won't solve the problem here. You'll need more sophisticated analysis than that.

**Comment by BraisGabin** [comment]
🤔 why not?

**Comment by schalkms** [comment]
I didn't know that we already have dataflow analysis as part of the `bindingContext`. Great. Thanks for bringing this up.
This issue can be reopened.

https://github.com/detekt/detekt/blob/b1269b58dc704c096ed402f91c9962942989a18a/detekt-rules-errorprone/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/NullableToStringCall.kt#L66-L78

**Comment by schalkms** [comment]
By the way, this has nothing to do with smart casts per se. The enabler here is data flow analysis as part of the Kotlin compiler. With the help of data flow analysis we can detect whether null checks or the necessary casts have been made before the `when` clause.
